plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ImmutableProjectState.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2020 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ide.xtext.server.build;
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.n4js.xtext.server.LSPIssue;
+import org.eclipse.xtext.resource.IResourceDescriptions;
+import org.eclipse.xtext.resource.impl.ResourceDescriptionsData;
+
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ListMultimap;
+
+/**
+ * Represents a consistent snapshot of the project state after a build was run.
+ */
+@SuppressWarnings("deprecation")
+public class ImmutableProjectState {
+	/** Type index */
+	private final ResourceDescriptionsData indexData;
+	/** File Mappings */
+	private final XSource2GeneratedMapping fileMappings;
+	/** Hashes to indicate file changes */
+	private final ImmutableMap<URI, HashedFileContent> fileHashes;
+	/** Hashes to indicate file changes */
+	private final ImmutableListMultimap<URI, LSPIssue> validationIssues;
+
+	/**
+	 * Factory method to make it explicit that we create copies of the passed arguments.
+	 */
+	public static ImmutableProjectState copyFrom(
+			ResourceDescriptionsData indexData,
+			XSource2GeneratedMapping fileMappings,
+			Map<? extends URI, ? extends HashedFileContent> fileHashes,
+			ListMultimap<? extends URI, ? extends LSPIssue> validationIssues) {
+		return new ImmutableProjectState(indexData.copy(), fileMappings.copy(), ImmutableMap.copyOf(fileHashes),
+				ImmutableListMultimap.copyOf(validationIssues));
+	}
+
+	/**
+	 * Returns an empty project state. No index data, no issues, no file mappings, nothing.
+	 */
+	public static ImmutableProjectState empty() {
+		return new ImmutableProjectState(new ResourceDescriptionsData(Collections.emptyList()),
+				new XSource2GeneratedMapping(), ImmutableMap.of(), ImmutableListMultimap.of());
+	}
+
+	/**
+	 * Use with caution.
+	 */
+	static ImmutableProjectState withoutCopy(
+			ResourceDescriptionsData indexData,
+			XSource2GeneratedMapping fileMappings,
+			ImmutableMap<URI, HashedFileContent> fileHashes,
+			ImmutableListMultimap<URI, LSPIssue> validationIssues) {
+		return new ImmutableProjectState(indexData, fileMappings, fileHashes, validationIssues);
+	}
+
+	/** Non-copying constructor */
+	private ImmutableProjectState(
```

Comment:
In order to guarantee immutability, these classes will likely be final eventually (I hope). Otherwise I'm afraid that we'll make our clients suffer from hard to track down bugs.